[pytorch] reintroduce static dispatch

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51957 [pytorch] reintroduce static dispatch**

This is a simplified version of #51554.

Compared to #51554, this version only supports statically dispatching to
a specific backend. The benefit is that it skipped the dispatch key
computation logic thus has less framework overhead. The downside is that
if input tensors do not match the specified backend it will throw error
instead of falling back to regular dispatch.

Sample code:
```
Tensor empty(IntArrayRef size, TensorOptions options, c10::optional<MemoryFormat> memory_format) {
    return at::cpu::empty(size, options, memory_format);
}

// aten::conj(Tensor(a) self) -> Tensor(a)
Tensor conj(const Tensor & self) {
    return at::math::conj(self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_out(Tensor & out, const Tensor & self) {
    return at::cpu::conj_out(out, self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_outf(const Tensor & self, Tensor & out) {
    return at::cpu::conj_out(out, self);
}

// aten::_conj(Tensor self) -> Tensor
Tensor _conj(const Tensor & self) {
    return at::defaultbackend::_conj(self);
}
```

For ops without the specific backend dispatch, it will throw error:
```
// aten::_use_cudnn_ctc_loss(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank) -> bool
bool _use_cudnn_ctc_loss(const Tensor & log_probs, const Tensor & targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t blank) {
    TORCH_CHECK(false, "Static dispatch does not support _use_cudnn_ctc_loss for CPU.");
}
```

Differential Revision: [D26337857](https://our.internmc.facebook.com/intern/diff/D26337857)

ljk53 committed:
[pytorch] reintroduce static dispatch

This is a simplified version of #51554.

Compared to #51554, this version only supports statically dispatching to
a specific backend. The benefit is that it skipped the dispatch key
computation logic thus has less framework overhead. The downside is that
if input tensors do not match the specified backend it will throw error
instead of falling back to regular dispatch.

Sample code:
```
Tensor empty(IntArrayRef size, TensorOptions options, c10::optional<MemoryFormat> memory_format) {
    return at::cpu::empty(size, options, memory_format);
}

// aten::conj(Tensor(a) self) -> Tensor(a)
Tensor conj(const Tensor & self) {
    return at::math::conj(self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_out(Tensor & out, const Tensor & self) {
    return at::cpu::conj_out(out, self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_outf(const Tensor & self, Tensor & out) {
    return at::cpu::conj_out(out, self);
}

// aten::_conj(Tensor self) -> Tensor
Tensor _conj(const Tensor & self) {
    return at::defaultbackend::_conj(self);
}
```

For ops without the specific backend dispatch, it still uses the c10 dispatch, e.g.:
```
// aten::_use_cudnn_ctc_loss(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank) -> bool
bool _use_cudnn_ctc_loss(const Tensor & log_probs, const Tensor & targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t blank) {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::_use_cudnn_ctc_loss", "")
        .typed<bool (const Tensor &, const Tensor &, IntArrayRef, IntArrayRef, int64_t)>();
    return op.call(log_probs, targets, input_lengths, target_lengths, blank);
}
```

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] reintroduce static dispatch"


This is a simplified version of #51554.

Compared to #51554, this version only supports statically dispatching to
a specific backend. The benefit is that it skipped the dispatch key
computation logic thus has less framework overhead. The downside is that
if input tensors do not match the specified backend it will throw error
instead of falling back to regular dispatch.

Sample code:
```
Tensor empty(IntArrayRef size, TensorOptions options, c10::optional<MemoryFormat> memory_format) {
    return at::cpu::empty(size, options, memory_format);
}

// aten::conj(Tensor(a) self) -> Tensor(a)
Tensor conj(const Tensor & self) {
    return at::math::conj(self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_out(Tensor & out, const Tensor & self) {
    return at::cpu::conj_out(out, self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_outf(const Tensor & self, Tensor & out) {
    return at::cpu::conj_out(out, self);
}

// aten::_conj(Tensor self) -> Tensor
Tensor _conj(const Tensor & self) {
    return at::defaultbackend::_conj(self);
}
```

For ops without the specific backend dispatch, it still uses the c10 dispatch, e.g.:
```
// aten::_use_cudnn_ctc_loss(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank) -> bool
bool _use_cudnn_ctc_loss(const Tensor & log_probs, const Tensor & targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t blank) {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::_use_cudnn_ctc_loss", "")
        .typed<bool (const Tensor &, const Tensor &, IntArrayRef, IntArrayRef, int64_t)>();
    return op.call(log_probs, targets, input_lengths, target_lengths, blank);
}
```

Differential Revision: [D26337857](https://our.internmc.facebook.com/intern/diff/D26337857)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] reintroduce static dispatch"


This is a simplified version of #51554.

Compared to #51554, this version only supports statically dispatching to
a specific backend. The benefit is that it skipped the dispatch key
computation logic thus has less framework overhead. The downside is that
if input tensors do not match the specified backend it will throw error
instead of falling back to regular dispatch.

Sample code:
```
Tensor empty(IntArrayRef size, TensorOptions options, c10::optional<MemoryFormat> memory_format) {
    return at::cpu::empty(size, options, memory_format);
}

// aten::conj(Tensor(a) self) -> Tensor(a)
Tensor conj(const Tensor & self) {
    return at::math::conj(self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_out(Tensor & out, const Tensor & self) {
    return at::cpu::conj_out(out, self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_outf(const Tensor & self, Tensor & out) {
    return at::cpu::conj_out(out, self);
}

// aten::_conj(Tensor self) -> Tensor
Tensor _conj(const Tensor & self) {
    return at::defaultbackend::_conj(self);
}
```

For ops without the specific backend dispatch, it still uses the c10 dispatch, e.g.:
```
// aten::_use_cudnn_ctc_loss(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank) -> bool
bool _use_cudnn_ctc_loss(const Tensor & log_probs, const Tensor & targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t blank) {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::_use_cudnn_ctc_loss", "")
        .typed<bool (const Tensor &, const Tensor &, IntArrayRef, IntArrayRef, int64_t)>();
    return op.call(log_probs, targets, input_lengths, target_lengths, blank);
}
```

Differential Revision: [D26337857](https://our.internmc.facebook.com/intern/diff/D26337857)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] reintroduce static dispatch"


This is a simplified version of #51554.

Compared to #51554, this version only supports statically dispatching to
a specific backend. The benefit is that it skipped the dispatch key
computation logic thus has less framework overhead. The downside is that
if input tensors do not match the specified backend it will throw error
instead of falling back to regular dispatch.

Sample code:
```
Tensor empty(IntArrayRef size, TensorOptions options, c10::optional<MemoryFormat> memory_format) {
    return at::cpu::empty(size, options, memory_format);
}

// aten::conj(Tensor(a) self) -> Tensor(a)
Tensor conj(const Tensor & self) {
    return at::math::conj(self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_out(Tensor & out, const Tensor & self) {
    return at::cpu::conj_out(out, self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_outf(const Tensor & self, Tensor & out) {
    return at::cpu::conj_out(out, self);
}

// aten::_conj(Tensor self) -> Tensor
Tensor _conj(const Tensor & self) {
    return at::defaultbackend::_conj(self);
}
```

For ops without the specific backend dispatch, it still uses the c10 dispatch, e.g.:
```
// aten::_use_cudnn_ctc_loss(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank) -> bool
bool _use_cudnn_ctc_loss(const Tensor & log_probs, const Tensor & targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t blank) {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::_use_cudnn_ctc_loss", "")
        .typed<bool (const Tensor &, const Tensor &, IntArrayRef, IntArrayRef, int64_t)>();
    return op.call(log_probs, targets, input_lengths, target_lengths, blank);
}
```

Differential Revision: [D26337857](https://our.internmc.facebook.com/intern/diff/D26337857)

[ghstack-poisoned]

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -104,6 +104,38 @@ def cpp_string(s: str) -> str:
+    if not backends or f.manual_kernel_registration:
+        return None
+
+    assert len(backends) <= 1, "Multi-backend static dispatch is not yet supported!"
```
just curious, is there something preventing multi-backend static dispatch from working? It would probably require extra CI jobs, so I could imagine not bothering with it until there's a need for it.

bdhirsh replied:
oh yeah, nvm. I guess only certain combinations would be valid (CPU and QuantizedCPU) and there's not much point sorting through it unless there's a use case

bhosmer replied:
@bdhirsh I think the biggest reason to support single-backend dispatch is that only under that constraint can we completely dispense with dispatch key computation, as opposed to hoisting it out of the dispatcher into the generated wrappers. 

It's true that we may get use cases for the multi-key version of this, but I don't know that we need to build it out now - and if/when it happens, I think we'd definitely want to preserve the single-key version as a distinct codegen path.

bdhirsh commented on test/mobile/custom_build/build.sh:
```diff
@@ -54,6 +57,22 @@ run_default_build() {
+  BUILD_ROOT="${LIBTORCH_BUILD_ROOT}" \
+    "${SRC_ROOT}/scripts/build_mobile.sh" \
+    -DCMAKE_CXX_FLAGS="-DSTRIP_ERROR_MESSAGES" \
+    -DSTATIC_DISPATCH_BACKENDS=CPU \
```
Do you think that testing CPU static dispatch in CI is representative enough that we don't need to bother separately testing static dispatch for other backends? Maybe it is- if not though, we might want to either add tests for other backends, or enforce somewhere (e.g. codegen) that only CPU is allowed for now, until a use case for cuda static dispatch shows up somewhere.

bhosmer replied:
That's an idea - gate the codegen so it's only available (and tested) on the backends that current customers need. When somebody asks for another backend, we enable it and add tests which use that customer's build setup :)

bhosmer commented on tools/codegen/gen.py:
```diff
@@ -104,6 +104,38 @@ def cpp_string(s: str) -> str:
+        # TODO: for ops with structured_delegate it should check the dispatch table of
+        # the out variant instead. For now, these structured ops all have CPU/CUDA kernels
+        # so we always dispatch to the target_backend, but this could be wrong when we
+        # migrate math/default_backend ops to use structured delegate.
```
Per our chat about emitting a warning in the TODO situation - if it's easy I'd throw it in (it seems like just an extra check based on the info you already have available here, right?)

ljk53 replied:
I thought about it again - I'm a little concerned it will generate a lot warnings when we start migrating more CPU ops to use structure kernels. The TODO situation should not be harmful as long as it compiles, i.e. when the structured-delegate actually has the CPU kernel it should be correct to call at::cpu::delegate_op(). If it doesn't have CPU kernel, then the build will fail anyway - we cannot print warning for the failure case unless we actually query the structured_delegate's dispatch table (basically what TODO needs to do).

I moved this TODO out of the for loop to make it more clear.

bhosmer reviewed (approved):
🚀

bhosmer commented:
Oh, one other thing - might we want to actually call it "single key static dispatch" or something similar? I.e. to avoid people assuming it's a straight substitute for dynamic dispatch / reincarnation of the old static dispatch.

ljk53 commented:
> Oh, one other thing - might we want to actually call it "single key static dispatch" or something similar? I.e. to avoid people assuming it's a straight substitute for dynamic dispatch / reincarnation of the old static dispatch.

@bhosmer Shall we call it "single key" or "single backend"? a runtime dispatch key (CPU) can dispatch to different registered keys (CPU, DefaultBackend, Math) - I'm not sure how to deal with the subtleties...

Do you suggest also changing all data structures from 'List[DispatchKey]' to 'DispatchKey' instead of asserting 'len(backends) <= 1'?

bhosmer commented:
> @bhosmer Shall we call it "single key" or "single backend"? a runtime dispatch key (CPU) can dispatch to different registered keys (CPU, DefaultBackend, Math) - I'm not sure how to deal with the subtleties...
> 
> Do you suggest also changing all data structures from 'List[DispatchKey]' to 'DispatchKey' instead of asserting 'len(backends) <= 1'?

Ah yeah you're totally right. Single key is wrong, single backend is right. 

Re data structure change - I think it depends on how permanent we consider the single-backend restriction. If we anticipated never getting a call for multi-backend static dispatch, then yeah (and also the command-line flag). But personally I'd leave it as-is in anticipation of eventually supporting multi-backend. Your call though. 

Oh that reminds me of one other question, re generating regular dispatched wrappers for ops that don't have kernels defined for the specified backend: wouldn't we want to just omit those ops entirely?

ljk53 committed:
Update on "[pytorch] reintroduce static dispatch"


This is a simplified version of #51554.

Compared to #51554, this version only supports statically dispatching to
a specific backend. The benefit is that it skipped the dispatch key
computation logic thus has less framework overhead. The downside is that
if input tensors do not match the specified backend it will throw error
instead of falling back to regular dispatch.

Sample code:
```
Tensor empty(IntArrayRef size, TensorOptions options, c10::optional<MemoryFormat> memory_format) {
    return at::cpu::empty(size, options, memory_format);
}

// aten::conj(Tensor(a) self) -> Tensor(a)
Tensor conj(const Tensor & self) {
    return at::math::conj(self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_out(Tensor & out, const Tensor & self) {
    return at::cpu::conj_out(out, self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_outf(const Tensor & self, Tensor & out) {
    return at::cpu::conj_out(out, self);
}

// aten::_conj(Tensor self) -> Tensor
Tensor _conj(const Tensor & self) {
    return at::defaultbackend::_conj(self);
}
```

For ops without the specific backend dispatch, it still uses the c10 dispatch, e.g.:
```
// aten::_use_cudnn_ctc_loss(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank) -> bool
bool _use_cudnn_ctc_loss(const Tensor & log_probs, const Tensor & targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t blank) {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::_use_cudnn_ctc_loss", "")
        .typed<bool (const Tensor &, const Tensor &, IntArrayRef, IntArrayRef, int64_t)>();
    return op.call(log_probs, targets, input_lengths, target_lengths, blank);
}
```

Differential Revision: [D26337857](https://our.internmc.facebook.com/intern/diff/D26337857)

[ghstack-poisoned]

ljk53 commented:
> Ah yeah you're totally right. Single key is wrong, single backend is right.
> Re data structure change - I think it depends on how permanent we consider the single-backend restriction. If we anticipated never getting a call for multi-backend static dispatch, then yeah (and also the command-line flag). But personally I'd leave it as-is in anticipation of eventually supporting multi-backend. Your call though.

Ok, I changed the codegen flag to make it clear it's for single backend. Also changed all "static_dispatch_backends" to "static_dispatch_backend" as well as the data structure. Just to keep them consistent.

> Oh that reminds me of one other question, re generating regular dispatched wrappers for ops that don't have kernels defined for the specified backend: wouldn't we want to just omit those ops entirely?

Ok, I'll generate `TORCH_CHECK(false, "Static dispatch does not support op XYZ for CPU backend");` for these ops.

ljk53 committed:
Update on "[pytorch] reintroduce static dispatch"


This is a simplified version of #51554.

Compared to #51554, this version only supports statically dispatching to
a specific backend. The benefit is that it skipped the dispatch key
computation logic thus has less framework overhead. The downside is that
if input tensors do not match the specified backend it will throw error
instead of falling back to regular dispatch.

Sample code:
```
Tensor empty(IntArrayRef size, TensorOptions options, c10::optional<MemoryFormat> memory_format) {
    return at::cpu::empty(size, options, memory_format);
}

// aten::conj(Tensor(a) self) -> Tensor(a)
Tensor conj(const Tensor & self) {
    return at::math::conj(self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_out(Tensor & out, const Tensor & self) {
    return at::cpu::conj_out(out, self);
}

// aten::conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
Tensor & conj_outf(const Tensor & self, Tensor & out) {
    return at::cpu::conj_out(out, self);
}

// aten::_conj(Tensor self) -> Tensor
Tensor _conj(const Tensor & self) {
    return at::defaultbackend::_conj(self);
}
```

For ops without the specific backend dispatch, it will throw error:
```
// aten::_use_cudnn_ctc_loss(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank) -> bool
bool _use_cudnn_ctc_loss(const Tensor & log_probs, const Tensor & targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t blank) {
    TORCH_CHECK(false, "Static dispatch does not support _use_cudnn_ctc_loss for CPU.");
}
```

Differential Revision: [D26337857](https://our.internmc.facebook.com/intern/diff/D26337857)

[ghstack-poisoned]
